overlord/devicestate/devicestate.go

```diff
@@ -585,19 +649,25 @@ func (r *remodeler) maybeInstallOrUpdate(ctx context.Context, st *state.State, r
 		}
 
 		return remodelChannelSwitch, []*state.TaskSet{ts}, nil
+	case needsComponentChanges:
+		tss, err := r.installComponents(ctx, st, currentInfo, requiredComponents)
+		if err != nil {
+			return 0, nil, err
+		}
+		return remodelAddComponentsAction, tss, nil
 	default:
 		// nothing to do but add the snap to the prereq tracker
 		r.tracker.Add(currentInfo)
 		return remodelNoAction, nil, nil
 	}
 }
 
-func (r *remodeler) shouldJustSwitch(rt remodelSnapTarget, needsRevisionChange bool) bool {
+func (r *remodeler) shouldSwitchWithoutRefresh(rt remodelSnapTarget, needsRevisionChange bool, needsComponentChanges bool) bool {
```

Comment:
I'm not entirely sure if passing needsComponentChanges is more readable than checking it outside of the helper